File: WebClient SDK/src/default/ui/krProcess/krStageTemplateUIExtension.ts
import { observe, Lambda } from 'mobx';
import { CardUIExtension, ICardUIExtensionContext } from 'tessa/ui/cards';
import { GridViewModel, LabelViewModel, AutoCompleteTableViewModel } from 'tessa/ui/cards/controls';
import { Card, CardRow, CardRowState } from 'tessa/cards';
import { LocalizationManager } from 'tessa/localization';
import { Visibility, DotNetType, Guid } from 'tessa/platform';
import { ValidationResultType } from 'tessa/platform/validation';

export class KrStageTemplateUIExtension extends CardUIExtension {
  public initialized(context: ICardUIExtensionContext) {
    const cardModel = context.model;

    if (cardModel.card.typeId !== '2fa85bb3-bba4-4ab6-ba97-652106db96de') { // KrStageTemplateTypeID
      return;
    }

    const grid = cardModel.controls.get('ApprovalStagesTable') as GridViewModel;
    if (!grid) {
      return;
    }

    let approversControlDisposer: Lambda | null = null;
    grid.rowInitializing.add(e => {
      const hyperlink = e.rowModel!.controls.get('AddComputedRoleLink') as LabelViewModel;
      if (!hyperlink) {
        return;
      }

      hyperlink.controlVisibility = KrStageTemplateUIExtension.hasComputedRole(
        e.cardModel.card,
        e.row.rowId
      )
        ? Visibility.Collapsed
        : Visibility.Visible;

      const approversControl = e.rowModel!.controls.get(
        'MultiplePerformersTableAC'
      ) as AutoCompleteTableViewModel;
      if (approversControl) {
        const approversControlChanged = () => {
          hyperlink.controlVisibility = KrStageTemplateUIExtension.hasComputedRole(
            e.cardModel.card,
            e.row.rowId
          )
            ? Visibility.Collapsed
            : Visibility.Visible;
        };

        approversControlDisposer = observe(approversControl, 'items', approversControlChanged);
      }

      const card = e.cardModel.card;
      const rows = card.sections.get('KrPerformersVirtual_Synthetic')!.rows;
      const rowId = e.row.rowId;

      hyperlink.onClick = () => {
        if (!KrStageTemplateUIExtension.hasComputedRole(card, rowId)) {
          const row = new CardRow();
          row.rowId = Guid.newGuid();
          row.state = CardRowState.Inserted;

          row.set('PerformerID', 'cd4d4a0d-414f-478d-a226-319aa8417f88', DotNetType.Guid); // SqlApproverRoleID
          row.set('PerformerName', '$KrProcess_SqlPerformersRole', DotNetType.String);
          row.set('StageRowID', rowId, DotNetType.Guid);
          const order = rows.length !== 0 ? Math.max(...rows.map(x => x.get('Order')), -1) + 1 : 0;
          row.set('Order', order, DotNetType.Int32);
          rows.push(row);
        }
      };
    });

    grid.rowEditorClosed.add(_ => {
      if (approversControlDisposer) {
        approversControlDisposer();
      }
    });

    grid.rowValidating.add(e => {
      const multiplePerformerControl = e.rowModel!.controls.get('MultiplePerformersTableAC');
      if (
        !multiplePerformerControl ||
        multiplePerformerControl.controlVisibility !== Visibility.Visible
      ) {
        return;
      }

      if (
        KrStageTemplateUIExtension.hasSqlQuery(e.row) &&
        e.row.isChanged('SqlApproverRole') &&
        !KrStageTemplateUIExtension.hasComputedRole(e.cardModel.card, e.row.rowId)
      ) {
        e.validationResult.add(
          ValidationResultType.Warning,
          LocalizationManager.instance.localize('$CardTypes_Validators_SQLApproversRoleMissed')
        );
      }
    });
  }

  private static hasSqlQuery(row: CardRow): boolean {
    return !!row.get('SqlApproverRole');
  }

  private static hasComputedRole(card: Card, stageRowId: guid): boolean {
    const sectionAlias =
      card.typeId === '4a377758-2366-47e9-98ac-c5f553974236'
        ? 'KrPerformersVirtual'
        : 'KrPerformersVirtual_Synthetic';

    const section = card.sections.tryGet(sectionAlias);
    if (!section) {
      return false;
    }

    return section.rows.some(row =>
      KrStageTemplateUIExtension.hasComputedRoleInRow(row, stageRowId)
    );
  }

  private static hasComputedRoleInRow(row: CardRow, stageRowId: guid): boolean {
    const srid = row.get('StageRowID');
    if (srid !== stageRowId) {
      return false;
    }

    const approverId = row.get('PerformerID');
    if (approverId !== 'cd4d4a0d-414f-478d-a226-319aa8417f88') {
      // SqlApproverRoleID
      return false;
    }

    return row.state !== CardRowState.Deleted;
  }
}

File: Source/Extensions/Tessa.Extensions.Default.Client/UI/KrProcess/KrStageTemplateUIExtension.cs
using System;
using System.Collections.Specialized;
using System.Linq;
using System.Threading.Tasks;
using System.Windows;
using Tessa.Cards;
using Tessa.Extensions.Default.Shared.Workflow.KrProcess;
using Tessa.Localization;
using Tessa.Platform;
using Tessa.Platform.Storage;
using Tessa.Platform.Validation;
using Tessa.UI;
using Tessa.UI.Cards;
using Tessa.UI.Cards.Controls;
using Tessa.UI.Cards.Controls.AutoComplete;

namespace Tessa.Extensions.Default.Client.UI.KrProcess
{
    public sealed class KrStageTemplateUIExtension : CardUIExtension
    {
        #region constants

        private const string TableAlias = KrConstants.Ui.KrApprovalStagesControlAlias;

        private const string AddComputedRoleLink = KrConstants.Ui.AddComputedRoleLink;

        private const string Approvers = KrConstants.Ui.KrMultiplePerformersTableAcAlias;

        #endregion

        #region private

        private static bool HasComputedRole(Card card, Guid stageRowID) =>
            card.TryGetPerformersSection(out var sec)
            && sec.Rows.Any(row => HasComputedRoleInRow(row, stageRowID));

        private static bool HasComputedRoleInRow(CardRow row, Guid stageRowID) =>
            row.Fields.TryGetValue(KrConstants.KrPerformersVirtual.StageRowID, out var srid)
            && stageRowID.Equals(srid)
            && row.Fields.TryGetValue(KrConstants.KrPerformersVirtual.PerformerID, out var approverID)
            && KrConstants.SqlApproverRoleID.Equals(approverID)
            && row.State != CardRowState.Deleted;

        private static bool HasHiddenStages(
            Card card)
        {
            return card.TryGetStagesSection(out var sec, preferVirtual: true)
                && sec.Rows.Any(p => p.TryGet<bool?>(KrConstants.KrStages.Hidden) == true);
        }

        #endregion

        #region base overrides

        public override async Task Initialized(ICardUIExtensionContext context)
        {
            var cardModel = context.Model;

            if (cardModel.Controls.TryGet(KrConstants.Ui.CanMoveCheckboxAlias, out var canMove)
                && canMove is CheckBoxViewModel canMoveCheckbox)
            {
                canMoveCheckbox.PropertyChanged += (
                    s,
                    e) =>
                {
                    if (s is CheckBoxViewModel checkbox
                        && e.PropertyName == "IsChecked"
                        && checkbox.IsChecked
                        && HasHiddenStages(cardModel.Card))
                    {
                        TessaDialog.ShowNotEmpty(
                            new ValidationResultBuilder()
                                // Там есть вложенные строки локализации, которые формат может разобрать
                                .AddWarning(LocalizationManager.Format("$KrProcess_HiddenStageWillBeRevealed")));
                    }
                };
            }

            if (cardModel.Controls.TryGet(TableAlias, out var control)
                && control is GridViewModel grid)
            {
                NotifyCollectionChangedEventHandler approversControlChanged = null;

                grid.RowInitializing += (s, e) =>
                {
                    e.Window.Width *= 1.5;

                    if (e.RowModel.Controls.TryGet(KrConstants.Ui.KrHiddenStageCheckboxAlias, out var hiddenCheckbox))
                    {
                        var canChangeOrder = e.CardModel.Card.Sections.TryGetValue(KrConstants.KrStageTemplates.Name, out var section)
                            && (section.Fields.TryGet<bool?>(KrConstants.KrStageTemplates.CanChangeOrder) ?? default);
                        // Если шаблон можно перемещать, то контрол становится ридонли
                        hiddenCheckbox.IsReadOnly = canChangeOrder;
                    }

                    if (e.RowModel.Controls.TryGet(AddComputedRoleLink, out var hyperlinkControl)
                        && hyperlinkControl is HyperlinkViewModel hyperlink)
                    {
                        hyperlink.ControlVisibility = HasComputedRole(e.CardModel.Card, e.Row.RowID)
                            ? Visibility.Collapsed
                            : Visibility.Visible;

                        if (e.RowModel.Controls.TryGet(Approvers, out var approversControlObj)
                            && approversControlObj is AutoCompleteTableViewModel approversControl)
                        {
                            approversControlChanged = (s1, e1) =>
                            {
                                var oldVisibility = hyperlink.ControlVisibility;
                                hyperlink.ControlVisibility = HasComputedRole(e.CardModel.Card, e.Row.RowID)
                                    ? Visibility.Collapsed
                                    : Visibility.Visible;
                                if (oldVisibility != hyperlink.ControlVisibility)
                                {
                                    if (hyperlink.ControlVisibility == Visibility.Visible
                                        && hyperlink.Block.BlockVisibility == Visibility.Collapsed)
                                    {
                                        hyperlink.Block.Form.Rearrange();
                                    }
                                    else
                                    {
                                        hyperlink.Block.RearrangeSelf();
                                    }
                                }
                            };

                            approversControl.Items.CollectionChanged += approversControlChanged;
                        }

                        var card = e.CardModel.Card;
                        var rows = card.Sections[KrConstants.KrPerformersVirtual.Synthetic].Rows;
                        var rowID = e.Row.RowID;

                        hyperlink.CommandClosure.Execute = _ =>
                        {
                            if (!HasComputedRole(e.CardModel.Card, e.Row.RowID))
                            {
                                var row = new CardRow
                                {
                                    RowID = Guid.NewGuid(),
                                    State = CardRowState.Inserted
                                };
                                row.Fields[KrConstants.KrPerformersVirtual.PerformerID] = KrConstants.SqlApproverRoleID;
                                row.Fields[KrConstants.KrPerformersVirtual.PerformerName] = "$KrProcess_SqlPerformersRole";
                                row.Fields[KrConstants.KrPerformersVirtual.StageRowID] = rowID;
                                row.Fields[KrConstants.KrPerformersVirtual.Order] = Int32Boxes.Box(
                                    rows.Count > 0 
                                    ? rows.Max(p => p.Fields.Get<int>(KrConstants.KrPerformersVirtual.Order)) + 1
                                    : 0);
                                rows.Add(row);
                            }
                        };
                    }
                };

                grid.RowEditorClosed += (s, e) =>
                {
                    if (approversControlChanged != null
                        && e.RowModel.Controls.TryGet(Approvers, out var approversControlObj)
                        && approversControlObj is AutoCompleteTableViewModel approversControl)
                    {
                        approversControl.Items.CollectionChanged -= approversControlChanged;
                    }
                };
            }
            
        }

        #endregion
    }
}